src/gluonnlp/data/utils.py

```diff
@@ -261,14 +264,47 @@ def train_valid_split(dataset, valid_ratio=0.05):
     if not 0.0 <= valid_ratio <= 1.0:
         raise ValueError('valid_ratio should be in [0, 1]')
 
-    num_train = len(dataset)
-    num_valid = np.ceil(num_train * valid_ratio).astype('int')
-    indices = np.arange(num_train)
+    if not stratify:
+        num_train = len(dataset)
+        num_valid = np.ceil(num_train * valid_ratio).astype('int')
+        indices = np.arange(num_train)
 
-    np.random.shuffle(indices)
-    valid = SimpleDataset([dataset[indices[i]] for i in range(num_valid)])
-    train = SimpleDataset([dataset[indices[i + num_valid]] for i in range(num_train - num_valid)])
-    return train, valid
+        np.random.shuffle(indices)
+
+        valid = SimpleDataset([dataset[indices[i]] for i in range(num_valid)])
+        train = SimpleDataset(
+            [dataset[indices[i + num_valid]] for i in range(num_train - num_valid)])
+
+        return train, valid
+    else:
+        if not isinstance(stratify, list):
+            raise TypeError('stratify should be a list')
+        if not len(stratify) == len(dataset):
+            raise ValueError('stratify should be the same length as num_train')
+
+        classes, digitized = np.unique(stratify, return_inverse=True)
+        n_classes = len(classes)
+        num_class = np.bincount(digitized)
```

Comment:
One problem of using bincount is that `len(num_class) != n_classes` in some cases, e.g., labels = [0,1,2,4] in which 3 is missing.